### PR TITLE
Allow empty result of numerical operations between DataArrays

### PR DESCRIPTION
**Pull request by MaximilianR**
Resolves https://github.com/pydata/xarray/issues/739

@shoyer let me know your thoughts and I'll add what's new note

### CODE REVIEW

**Comment by MaximilianR** [comment]
Not sure if I'm doing something foolish - I had some issues running the tests on my machine because of the `from .version import version as __version__`. Let me know if there's a quick fix beyond me commenting that line out.

**Comment by MaximilianR** [comment]
Forgive the python question - please disregard unless you know the answer off hand - do you know if there any way of monkey patching this into our code? I did that for the 0.7 fixes we added, but I'm not sure that's possible here: `_binary_op` gets called many times upon `import xarray`, and so I think the prior `_binary_op` is already 'embedded' in all the functions before we have a chance to monkey patch.

**Comment by spencerahill** [comment]
@MaximilianR FWIW I had problems with `.version` recently when I cloned from the main repo...had to copy the `version.py` from another xarray install.  Maybe has something to do with .gitignore: xarray/version.py is included there.

**Comment by shoyer** [comment]
Version.py gets created when you install xarray, e.g., when you run `pip install -e .` after running git clone.

**Comment by spencerahill** [comment]
@shoyer thanks!  Good to know.

**Comment by shoyer** [comment]
For consistency, we should also fix arithmetic with Dataset (allow no overlapping labels and no overlapping variables) and GroupBy objects:
https://github.com/pydata/xarray/blob/v0.7.0/xarray/core/dataset.py#L1933
https://github.com/pydata/xarray/blob/v0.7.0/xarray/core/dataset.py#L1975
https://github.com/pydata/xarray/blob/v0.7.0/xarray/core/groupby.py#L220

**Comment by MaximilianR** [comment]
Super - those are now updated @shoyer

File: xarray/test/test_dataset.py [hunk 1]
```diff
@@ -1868,8 +1868,9 @@ def test_fillna(self):
         expected = Dataset({'a': ('x', [-1, 1, -1, 3]), 'b': np.nan})
         self.assertDatasetIdentical(expected, actual)
 
-        with self.assertRaisesRegexp(ValueError, 'no overlapping'):
-            ds.fillna({'x': 0})
+        actual=ds.fillna({'x': 0})
+        expected=ds
+        self.assertDatasetEqual(actual, expected)
```
Comment (MaximilianR) [inline xarray/test/test_dataset.py]: I think these may now be *too* liberal. Here, there is no variable called `x`. I think this _should_ raise - it seems fairly different to multiply empty arrays vs supplying variables that don't exist.

I'll ponder further, let me know if you have any thoughts on this in the meantime

**Comment by MaximilianR** [comment]
@shoyer I think this is a decent balance:
- Allow empty dimensions
- But raise if non-existent dimensions are supplied

I think the code and tests now reflect that. Let me know your thoughts / comments

**Comment by MaximilianR** [comment]
This is my last PR that's ready to go @shoyer

**Comment by shoyer** [comment]
Still thinking about the fillna behavior... I agree, it's weird that `ds.fillna({'missing_variable': 0})` works without raising an error.

I would say this is consistent with in-place arithmetic on Datasets (i.e., left alignment of variable names), but we actually don't allow that with different variables currently... probably because I wanted to avoid this headache :). I think I'll add a special case in to ensure that we raise an error if there are unexpected variables in the fillna argument.

**Comment by shoyer** [comment]
OK, going to merge this...

**Comment by shoyer** [comment]
I'm probably going to call this an "enhancement" rather than a bug fix, but otherwise this looks good... I'll fix up the what's new page shortly.

**Comment by MaximilianR** [comment]
Cheers!

**Pull request merged**